pvlib/shading.py

```diff
@@ -0,0 +1,140 @@
+    average masking angle as described in [1]_.
+
+    Parameters
+    ----------
+    surface_tilt : numeric
+        Panel tilt from horizontal [degrees].
+
+    gcr : float
+        The ground coverage ratio of the array [unitless].
+
+    Returns
+    ----------
+    mask_angle : numeric
+        Average angle from horizontal where diffuse light is blocked by the
+        preceding row [degrees].
+
+    See Also
+    --------
+    passias_sky_diffuse
+
+    Notes
+    -----
+    The pvlib-python authors believe that Eqn. 9 in [1]_ is incorrect.
+    Here we use an independent equation.  First, Eqn. 8 is non-dimensionalized
+    (recasting in terms of GCR):
+
+    .. math::
+
+        \psi(z') = \arctan \left [
+            \frac{(1 - z') \sin \beta}
+                 {\mathrm{GCR}^{-1} + (z' - 1) \cos \beta}
+        \right ]
+
+    Where :math:`GCR = B/C` and :math:`z' = z/B`. The average masking angle
+    :math:`\overline{\psi} = \int_0^1 \psi(z') \mathrm{d}z'` is then
+    evaluated symbolically using Maxima (using :math:`X = 1/\mathrm{GCR}`):
+
+    .. code-block:: none
+
+        load(scifac)    /* for the gcfac function */
+        assume(X>0, cos(beta)>0, cos(beta)-X<0);   /* X is 1/GCR */
+        gcfac(integrate(atan((1-z)*sin(beta)/(X+(z-1)*cos(beta))), z, 0, 1))
+
+    This yields the equation implemented by this function:
+
+    .. math::
+
+        \overline{\psi} = \
+            &-\frac{X}{2} \sin\beta \log | 2 X \cos\beta - (X^2 + 1)| \\
+            &+ (X \cos\beta - 1) \arctan \frac{X \cos\beta - 1}{X \sin\beta} \\
+            &+ (1 - X \cos\beta) \arctan \frac{\cos\beta}{\sin\beta} \\
+            &+ X \log X \sin\beta
+
+    The pvlib-python authors have validated this equation against numerical
+    integration of :math:`\overline{\psi} = \int_0^1 \psi(z') \mathrm{d}z'`.
+
+    References
+    ----------
+    .. [1] D. Passias and B. Källbäck, "Shading effects in rows of solar cell
+       panels", Solar Cells, Volume 11, Pages 281-291.  1984.
+       DOI: 10.1016/0379-6787(84)90017-6
+    """
+    # wrap it in an array so that division by zero is handled well
+    beta = np.radians(np.array(surface_tilt))
+    sin_b = np.sin(beta)
+    cos_b = np.cos(beta)
+    X = 1/gcr
+
+    with np.errstate(divide='ignore', invalid='ignore'):  # ignore beta=0
+        term1 = -X * sin_b * np.log(np.abs(2 * X * cos_b - (X**2 + 1))) / 2
+        term2 = (X * cos_b - 1) * np.arctan((X * cos_b - 1) / (X * sin_b))
+        term3 = (1 - X * cos_b) * np.arctan(cos_b / sin_b)
+        term4 = X * np.log(X) * sin_b
+
+    psi_avg = term1 + term2 + term3 + term4
+    # when beta=0, divide by zero makes psi_avg NaN.  replace with 0:
+    psi_avg = np.where(np.isfinite(psi_avg), psi_avg, 0)
+
+    if isinstance(surface_tilt, pd.Series):
+        psi_avg = pd.Series(psi_avg, index=surface_tilt.index)
+
+    return np.degrees(psi_avg)
+
+
+def passias_sky_diffuse(masking_angle):
```

Comment:
Wondering if this signature creates space for other approaches to calculate self-shading. What if the arguments are `tilt` and 'gcr' and `passias_masking_angle` was private and called within this function? I doubt that the average "masking angle" is general among other models, I could be wrong.

A thought: how is this visible sky dome calculation done in `bifacial_vf`? Maybe there's some commonality that could inform the interface here, although we wouldn't re-use code from `bifacial_vf`. If I recall right, that calculation is done with view factors over small increments of angle.

I have a mild preference for `sky_diffuse_passias` as the function name.

Reply:
I separated the two functions in anticipation of people wanting to use the "worst-case" angle like SAM does instead of Passias et al.'s average angle.  That said, a function that just does `1 - cos(angle/2)^2` does seem strange.  

Good idea about bifacial_vf; I haven't had an excuse to get familiar with it before now but am happy to have one.

Reply:
I looked at the SAM documentation. I think that the low level functions that calculates the average `masking_angle` is useful, as would be a function that returns `masking_angle` at a point along the row's surface. Maybe someone would want to use the `masking_angle` in combination with other sky diffuse models.

Still chewing on the name `passias_sky_diffuse` for the effects function.

Reply:
I added a `masking_angle` function that calculates the angle for a given point and renamed `passias_sky_diffuse` to `sky_diffuse_passias`, but am happy to rename if you have a better alternative.

> Maybe someone would want to use the masking_angle in combination with other sky diffuse models.

It occurs to me that the masking angle is also useful for comparing with solar position to determine if direct shading is relevant.